scripts/validate_response.py

```diff
@@ -0,0 +1,150 @@
+"""Response validation using pre-defined question/answer pair."""
+
+import argparse
+import json
+import os
+import sys
+from collections import defaultdict
+
+import requests
+from llama_index.embeddings.huggingface import HuggingFaceEmbedding
+from pandas import DataFrame
+from scipy.spatial.distance import cosine, euclidean
+
+
+# TODO: Add more question/answer pair
+def _args_parser(args):
+    """Arguments parser."""
+    parser = argparse.ArgumentParser(description="Response validation module.")
+    parser.add_argument(
+        "-s",
+        "--scenario",
+        choices=["with_rag", "without_rag"],
+        default="with_rag",
+        help="Scenario for which responses will be evaluated.",
+    )
+    parser.add_argument(
+        "-m",
+        "--model",
+        choices=["gpt", "granite"],
+        default="gpt",
+        help="Model for which responses will be evaluated.",
+    )
+    parser.add_argument(
+        "-t",
+        "--threshold",
+        type=lambda val: float(val),
+        default=0.25,
+        help="Threshold value to be used for similarity score.",
+    )
+    parser.add_argument(
+        "-n",
+        "--num_questions",
+        type=lambda val: int(val),
+        default=2,
+        help="Number of questions to be validated.",
+    )
+    return parser.parse_args(args)
+
+
+class ResponseValidation:
+    """Validate LLM response."""
+
+    def __init__(self):
+        """Initialize."""
+        self._embedding_model = HuggingFaceEmbedding("BAAI/bge-base-en")
+
+    def get_similarity_score(self, response, answer):
+        """Calculate similarity score between two strings."""
+        res_vec = self._embedding_model.get_text_embedding(response)
+        ans_vec = self._embedding_model.get_text_embedding(answer)
+
+        # Distance score
+        cos_score = cosine(res_vec, ans_vec)
+        euc_score = euclidean(res_vec, ans_vec)
+
+        # Naive length consideration with reduced weightage.
+        len_res, len_ans = len(response), len(answer)
+        len_score = (abs(len_res - len_ans) / (len_res + len_ans)) * 0.1
+
+        score = len_score + (cos_score + euc_score) / 2
+        # TODO: Consider both contextual/non-contextual embedding.
+
+        print(
+            f"cos_score: {cos_score}, "
+            f"euc_score: {euc_score}, "
+            f"len_score: {len_score}\n"
+            f"final_score: {score}"
+        )
+        return score
+
+    def get_response_quality(self, args, qa_pairs):
```

Comment:
Something to start with, this simple evaluation technique seems good. I hope we have a #TODO action item to look for frameworks like RAGAS etc, to do similar evaluations in future.